src/main/java/com/metamx/metrics/cgroups/ProcCgroupDiscoverer.java

```diff
@@ -0,0 +1,271 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.metamx.metrics.cgroups;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+import com.metamx.common.StringUtils;
+import com.metamx.common.logger.Logger;
+import com.metamx.metrics.CgroupUtil;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class ProcCgroupDiscoverer implements CgroupDiscoverer
+{
+  private static final Logger LOG = new Logger(ProcCgroupDiscoverer.class);
+  // TODO: discover `/proc` via whatever is mounted by `/proc`
+  private static final String CGROUP_TYPE = "cgroup";
+  private static final String PROC_TYPE = "proc";
+
+  @Override
+  public Path discover(final String cgroup, long pid)
+  {
+    Preconditions.checkNotNull(cgroup, "cgroup");
+    // Wish List: find a way to cache these
+    final File proc = proc();
+    if (proc == null) {
+      return null;
+    }
+    final File procMounts = getProcMounts(proc);
+    final File procCgroups = getProcCgroups(proc);
+    final File pidCgroups = getPidCgroups(proc, pid);
+    final ProcCgroupsEntry procCgroupsEntry = getCgroupEntry(procCgroups, cgroup);
+    final ProcMountsEntry procMountsEntry = getMountEntry(procMounts, cgroup);
+    final ProcPidCgroupEntry procPidCgroupEntry = getPidCgroupEntry(pidCgroups, procCgroupsEntry.hierarchy);
+    final File cgroupDir = new File(
+        procMountsEntry.path.toFile(),
+        procPidCgroupEntry.path
+    );
+    if (cgroupDir.exists() && cgroupDir.isDirectory()) {
+      return cgroupDir.toPath();
+    }
+    LOG.warn("Invalid cgroup directory [%s]", cgroupDir);
+    return null;
+  }
+
+  @VisibleForTesting
+  public long getProbabyPid()
+  {
+    try {
+      return CgroupUtil.getProbablyPID();
+    }
+    catch (CgroupUtil.IndeterminatePid indeterminatePid) {
+      throw Throwables.propagate(indeterminatePid);
+    }
+  }
+
+  @VisibleForTesting
+  public File proc()
+  {
+    final File proc = Paths.get("/proc").toFile();
+    Path foundProc = null;
+    if (proc.exists() && proc.isDirectory()) {
+      // Sanity check
+      try {
+        for (final String line : Files.readLines(new File(proc, "mounts"), Charsets.UTF_8)) {
```

Comment:
It starts looking at `/proc`. I couldn't find a reasonable way to programatically do this from java + files. Maybe allow a config to be passed in about where to find `/proc`?